### ISSUE TITLE
Restart failed after ls crash

### ISSUE THREAD

**Issue by lukaszsamson** [JakeBecker/vscode-elixir-ls#92]
version: 0.2.21

When ls crashed, the extension failed to restart it.
Possibly related to https://github.com/JakeBecker/elixir-ls/issues/98 as it happened right after it.

```
15:28:04.632 [info]  Application language_server exited: shutdown

[Warn  - 15:28:15] ** (exit) exited in: GenServer.call(ElixirLS.LanguageServer.JsonRpc, {:packet, %{"id" => 455, "jsonrpc" => "2.0", "method" => "textDocument/hover", "params" => %{"position" => %{"character" => 52, "line" => 526}, "textDocument" => %{"uri" => "file:///Users/user/my_app/apps/my_app/test/my_app/balance_test.exs"}}}}, 5000)
    ** (EXIT) no process: the process is not alive or there's no process currently associated with the given name, possibly because its application isn't started
    (elixir) lib/gen_server.ex:914: GenServer.call/3
    (elixir) lib/stream.ex:432: anonymous fn/4 in Stream.each/2
    (elixir) lib/stream.ex:1379: Stream.do_resource/5
    (elixir) lib/stream.ex:1553: Enumerable.Stream.do_each/4
    (elixir) lib/stream.ex:637: Stream.run/1
    (stdlib) erl_eval.erl:677: :erl_eval.do_apply/6
    (elixir) lib/code.ex:232: Code.eval_string/3

[Info  - 15:28:15] Connection to server got closed. Server will restart.
[Error - 15:28:15] Request textDocument/hover failed.
Error: Connection got disposed.
    at Object.dispose (/Users/user/.vscode/extensions/jakebecker.elixir-ls-0.2.21/node_modules/vscode-jsonrpc/lib/main.js:825:25)
    at Object.dispose (/Users/user/.vscode/extensions/jakebecker.elixir-ls-0.2.21/node_modules/vscode-languageclient/lib/client.js:57:35)
    at LanguageClient.handleConnectionClosed (/Users/user/.vscode/extensions/jakebecker.elixir-ls-0.2.21/node_modules/vscode-languageclient/lib/client.js:1864:38)
    at LanguageClient.handleConnectionClosed (/Users/user/.vscode/extensions/jakebecker.elixir-ls-0.2.21/node_modules/vscode-languageclient/lib/main.js:106:15)
    at closeHandler (/Users/user/.vscode/extensions/jakebecker.elixir-ls-0.2.21/node_modules/vscode-languageclient/lib/client.js:1852:18)
    at CallbackList.invoke (/Users/user/.vscode/extensions/jakebecker.elixir-ls-0.2.21/node_modules/vscode-jsonrpc/lib/events.js:71:39)
    at Emitter.fire (/Users/user/.vscode/extensions/jakebecker.elixir-ls-0.2.21/node_modules/vscode-jsonrpc/lib/events.js:135:36)
    at closeHandler (/Users/user/.vscode/extensions/jakebecker.elixir-ls-0.2.21/node_modules/vscode-jsonrpc/lib/main.js:221:26)
    at CallbackList.invoke (/Users/user/.vscode/extensions/jakebecker.elixir-ls-0.2.21/node_modules/vscode-jsonrpc/lib/events.js:71:39)
    at Emitter.fire (/Users/user/.vscode/extensions/jakebecker.elixir-ls-0.2.21/node_modules/vscode-jsonrpc/lib/events.js:135:36)
    at StreamMessageWriter.AbstractMessageWriter.fireClose (/Users/user/.vscode/extensions/jakebecker.elixir-ls-0.2.21/node_modules/vscode-jsonrpc/lib/messageWriter.js:57:27)
    at Socket.<anonymous> (/Users/user/.vscode/extensions/jakebecker.elixir-ls-0.2.21/node_modules/vscode-jsonrpc/lib/messageWriter.js:79:63)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:191:7)
    at Pipe._handle.close [as _onclose] (net.js:510:12)
```

Edit:
Happened another time, this time without `Error: Connection got disposed`.  Here is Debug console output
```
7console.ts:136 [Extension Host] Notification handler 'window/logMessage' failed with message: Channel has been closed
t.log @ console.ts:136
log.ts:171   ERR Channel has been closed: Error: Channel has been closed
	at e.validate (/Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/node/extensionHostProcess.js:583:372)
	at e.appendLine (/Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/node/extensionHostProcess.js:583:6)
	at LanguageClient.error (/Users/user/.vscode/extensions/jakebecker.elixir-ls-0.2.21/node_modules/vscode-languageclient/lib/client.js:1592:28)
	at LanguageClient.logFailedRequest (/Users/user/.vscode/extensions/jakebecker.elixir-ls-0.2.21/node_modules/vscode-languageclient/lib/client.js:2050:14)
	at client.sendRequest.then (/Users/user/.vscode/extensions/jakebecker.elixir-ls-0.2.21/node_modules/vscode-languageclient/lib/client.js:721:24)
	at <anonymous>
2console.ts:136 [Extension Host] Notification handler 'window/logMessage' failed with message: Channel has been closed
t.log @ console.ts:136
notificationsAlerts.ts:42 ElixirLS has crashed. See Output panel.
e.onDidNotificationChange @ notificationsAlerts.ts:42
2console.ts:136 [Extension Host] Notification handler 'window/logMessage' failed with message: Channel has been closed
t.log @ console.ts:136
console.ts:136 [Extension Host] Error: Channel has been closed
	at e.validate (/Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/node/extensionHostProcess.js:583:372)
	at e.appendLine (/Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/node/extensionHostProcess.js:583:6)
	at LanguageClient.info (/Users/user/.vscode/extensions/jakebecker.elixir-ls-0.2.21/node_modules/vscode-languageclient/lib/client.js:1574:28)
	at LanguageClient.handleConnectionClosed (/Users/user/.vscode/extensions/jakebecker.elixir-ls-0.2.21/node_modules/vscode-languageclient/lib/client.js:1875:18)
	at LanguageClient.handleConnectionClosed (/Users/user/.vscode/extensions/jakebecker.elixir-ls-0.2.21/node_modules/vscode-languageclient/lib/main.js:106:15)
	at closeHandler (/Users/user/.vscode/extensions/jakebecker.elixir-ls-0.2.21/node_modules/vscode-languageclient/lib/client.js:1852:18)
	at CallbackList.invoke (/Users/user/.vscode/extensions/jakebecker.elixir-ls-0.2.21/node_modules/vscode-jsonrpc/lib/events.js:71:39)
	at Emitter.fire (/Users/user/.vscode/extensions/jakebecker.elixir-ls-0.2.21/node_modules/vscode-jsonrpc/lib/events.js:135:36)
	at closeHandler (/Users/user/.vscode/extensions/jakebecker.elixir-ls-0.2.21/node_modules/vscode-jsonrpc/lib/main.js:221:26)
	at CallbackList.invoke (/Users/user/.vscode/extensions/jakebecker.elixir-ls-0.2.21/node_modules/vscode-jsonrpc/lib/events.js:71:39)
	at Emitter.fire (/Users/user/.vscode/extensions/jakebecker.elixir-ls-0.2.21/node_modules/vscode-jsonrpc/lib/events.js:135:36)
	at StreamMessageReader.AbstractMessageReader.fireClose (/Users/user/.vscode/extensions/jakebecker.elixir-ls-0.2.21/node_modules/vscode-jsonrpc/lib/messageReader.js:135:27)
	at Socket.<anonymous> (/Users/user/.vscode/extensions/jakebecker.elixir-ls-0.2.21/node_modules/vscode-jsonrpc/lib/messageReader.js:188:62)
	at emitOne (events.js:101:20)
	at Socket.emit (events.js:191:7)
	at Pipe._handle.close [as _onclose] (net.js:510:12)
t.log @ console.ts:136
log.ts:171   ERR Language client is not ready yet: Error: Language client is not ready yet
	at LanguageClient.sendRequest (/Users/user/.vscode/extensions/jakebecker.elixir-ls-0.2.21/node_modules/vscode-languageclient/lib/client.js:1473:19)
	at provideHover (/Users/user/.vscode/extensions/jakebecker.elixir-ls-0.2.21/node_modules/vscode-languageclient/lib/client.js:720:27)
	at Object.provideHover (/Users/user/.vscode/extensions/jakebecker.elixir-ls-0.2.21/node_modules/vscode-languageclient/lib/client.js:730:23)
	at /Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/node/extensionHostProcess.js:720:345
	at /Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/node/extensionHostProcess.js:109:515
	at new n.Class.derive._oncancel (/Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/node/extensionHostProcess.js:84:560)
	at Object.t.asWinJsPromise (/Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/node/extensionHostProcess.js:109:478)
	at e.provideHover (/Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/node/extensionHostProcess.js:720:299)
	at /Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/node/extensionHostProcess.js:734:302
	at e._withAdapter (/Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/node/extensionHostProcess.js:731:534)
	at e.$provideHover (/Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/node/extensionHostProcess.js:734:264)
	at e._doInvokeHandler (/Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/node/extensionHostProcess.js:658:87)
	at e._invokeHandler (/Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/node/extensionHostProcess.js:657:779)
	at e._receiveRequest (/Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/node/extensionHostProcess.js:656:851)
	at e._receiveOneMessage (/Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/node/extensionHostProcess.js:656:620)
	at /Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/node/extensionHostProcess.js:655:544
	at /Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/node/extensionHostProcess.js:658:673
	at /Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/node/extensionHostProcess.js:96:599
	at e.fire (/Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/node/extensionHostProcess.js:97:917)
	at u (/Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/node/extensionHostProcess.js:159:721)
	at Socket.<anonymous> (/Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/node/extensionHostProcess.js:159:898)
	at emitOne (events.js:96:13)
	at Socket.emit (events.js:191:7)
	at readableAddChunk (_stream_readable.js:178:18)
	at Socket.Readable.push (_stream_readable.js:136:10)
	at Pipe.onread (net.js:560:20)
```
The error messages repeat many times.
Process explorer on the other hand shows that the process is still running.
```
CPU %	Memory (MB)	pid	Name
0	98	15011	code main
1	66	15012	     gpu-process
0	49	15024	     shared-process
0	246	57837	     window (Untitled-1 — my_app)
0	98	57838	         extensionHost
0	33	57889	             electron_node server.js 
0	0	58225	             sh /Users/user/.vscode/extensions/jakebecker.elixir-ls-0.2.21/elixir-ls-release/language_server.sh
0	33	58233	                 /usr/local/Cellar/erlang/21.0.3/lib/erlang/erts-10.0.3/bin/beam.smp -- -root /usr/local/Cellar/erlang/21.0.3/lib/erlang -progname erl -- -home /Users/user -- -pa /Users/user/elixir/bin/../lib/eex/ebin /Users/user/elixir/bin/../lib/elixir/ebin /Users/user/elixir/bin/../lib/ex_unit/ebin /Users/user/elixir/bin/../lib/iex/ebin /Users/user/elixir/bin/../lib/logger/ebin /Users/user/elixir/bin/../lib/mix/ebin -noshell -s elixir start_cli -extra -e ElixirLS.LanguageServer.CLI.main()
0	0	58249	                     erl_child_setup 10496
0	66	57839	         watcherService
0	49	57871	         searchService
0	82	58665	     window (Process Explorer)
0	0	59171	         /bin/ps -ax -o pid=,ppid=,pcpu=,pmem=,command=
```

**Comment by lukaszsamson** [comment]
It looks like the `Object.dispose` error is already fixed upstream in `vscode-languageclient` >= 4.0.0 
https://github.com/Microsoft/vscode-languageserver-node/issues/286
and `Channel has been closed` handling was improoved in `vscode-languageclient` >= 4.1.1
https://github.com/Microsoft/vscode-languageserver-node/issues/319

**Comment by JakeBecker** [comment]
Good catch on the vscode-languageclient dependency issue! I updated it myself before noticing your PR. I've reverted to killing the VM when the server crashes, and that plus the dependency upgrade should fix it.
